connect/mirror/src/test/java/org/apache/kafka/connect/mirror/OffsetSyncStoreTest.java

```diff
@@ -57,52 +58,170 @@ public void testOffsetTranslation() {
+                    // Invariant D: the earliest sync is present
+                    assertEquals(0L, store.syncFor(tp, 63).upstreamOffset());
+                    int count = countDistinctStoredSyncs(store, tp);
+                    int diff = count - lastCount;
+                    assertTrue(diff >= -1,
+                            "Store expired too many syncs: " + diff + " after receiving offset " + offset);
+                    lastCount = count;
+                }
+            }
+        }
+    }
+
+    private void assertSparseSync(FakeOffsetSyncStore store, long syncOffset, long previousOffset) {
+        assertEquals(OptionalLong.of(previousOffset == -1 ? previousOffset : previousOffset + 1), store.translateDownstream(null, tp, syncOffset - 1));
+        assertEquals(OptionalLong.of(syncOffset), store.translateDownstream(null, tp, syncOffset));
+        assertEquals(OptionalLong.of(syncOffset + 1), store.translateDownstream(null, tp, syncOffset + 1));
+        assertEquals(OptionalLong.of(syncOffset + 1), store.translateDownstream(null, tp, syncOffset + 2));
+    }
+
+    private int countDistinctStoredSyncs(FakeOffsetSyncStore store, TopicPartition topicPartition) {
+        int count = 1;
+        for (int i = 1; i < OffsetSyncStore.SYNCS_PER_PARTITION; i++) {
+            if (store.syncFor(topicPartition, i - 1) != store.syncFor(topicPartition, i)) {
+                count++;
+            }
+        }
+        return count;
+    }
+
+    private void assertSparseSyncInvariant(FakeOffsetSyncStore store, TopicPartition topicPartition) {
+        for (int j = 0; j < OffsetSyncStore.SYNCS_PER_PARTITION; j++) {
+            for (int i = 0; i <= j; i++) {
+                long jUpstream = store.syncFor(topicPartition, j).upstreamOffset();
+                long iUpstream = store.syncFor(topicPartition, i).upstreamOffset();
+                if (jUpstream == iUpstream) {
+                    continue;
+                }
+                long iUpstreamLowerBound = jUpstream + (1L << Math.max(i - 2, 0));
+                if (iUpstreamLowerBound < 0) {
+                    continue;
+                }
+                assertTrue(
+                        iUpstreamLowerBound <= iUpstream,
+                        "Invariant C(" + i + "," + j + "): Upstream offset " + iUpstream + " at position " + i
+                                + " should be at least " + iUpstreamLowerBound + " (" + jUpstream + " + 2^" + Math.max(i - 1, 0) + ")"
+                );
+                long iUpstreamBound = jUpstream + (1L << j) - (1L << i);
+                if (iUpstreamBound < 0)
+                    continue;
+                assertTrue(
+                        iUpstream <= iUpstreamBound,
+                        "Invariant B(" + i + "," + j + "): Upstream offset " + iUpstream + " at position " + i
+                                + " should be no greater than " + iUpstreamBound + " (" + jUpstream + " + 2^" + j + ")"
```

Comment:
This doesn't match the invariant (or the right-hand expression for the declaration of `iUpstreamBound`):
```suggestion
                                + " should be no greater than " + iUpstreamBound + " (" + jUpstream + " + 2^" + j + " - 2^" + i + ")"
```